contracts/src/token/erc20/extensions/flash_mint.rs

```diff
@@ -0,0 +1,318 @@
+    ///
+    /// The fee is a fixed percentage of the borrowed amount.
+    ///
+    /// If the token is not supported, the function returns an
+    /// `UnsupportedToken` error.
+    ///
+    /// * `token` - The address of the ERC-20 token that will be loaned.
+    /// * `amount` - The amount of tokens that will be loaned.
+    fn flash_fee(
+        &self,
+        token: Address,
+        amount: U256,
+    ) -> Result<U256, Self::Error>;
+
+    /// Executes a flash loan.
+    ///
+    /// This function is part of the ERC-3156 (Flash Loans) standard.
+    ///
+    /// * `receiver` - The contract that will receive the flash loan.
+    /// * `token` - The ERC-20 token that will be loaned.
+    /// * `amount` - The amount of tokens that will be loaned.
+    /// * `data` - Arbitrary data that can be passed to the receiver contract.
+    ///
+    /// The function must return `true` if the flash loan was successful,
+    /// and revert otherwise.
+    fn flash_loan(
+        &mut self,
+        receiver: Address,
+        token: Address,
+        amount: U256,
+        data: Bytes,
+        erc20: &mut Erc20,
+    ) -> Result<bool, Self::Error>;
+}
+
+const RETURN_VALUE: [u8; 32] = keccak_const::Keccak256::new()
+    .update("ERC3156FlashBorrower.onFlashLoan".as_bytes())
+    .finalize();
+
+impl IERC3156FlashLender for Erc20FlashMint {
+    type Error = Error;
+
+    fn max_flash_loan(&self, token: Address, erc20: &Erc20) -> U256 {
+        if token == contract::address() {
+            return U256::MAX - erc20.total_supply();
+        }
+        U256::MIN
+    }
+
+    fn flash_fee(
+        &self,
+        token: Address,
+        _amount: U256,
+    ) -> Result<U256, Self::Error> {
+        if token != contract::address() {
+            return Err(Error::UnsupportedToken(ERC3156UnsupportedToken {
+                token,
+            }));
+        }
+        Ok(self.flash_fee_amount.get())
+    }
+
+    fn flash_loan(
+        &mut self,
+        receiver: Address,
+        token: Address,
+        value: U256,
+        data: Bytes,
+        erc20: &mut Erc20,
+    ) -> Result<bool, Self::Error> {
+        let max_loan = self.max_flash_loan(token, erc20);
+        if value > max_loan {
+            return Err(Error::ExceededMaxLoan(ERC3156ExceededMaxLoan {
+                max_loan,
+            }));
+        }
+
+        let fee = self.flash_fee(token, value)?;
+        erc20._mint(receiver, value)?;
+        let loan_receiver = IERC3156FlashBorrower::new(receiver);
+        if Address::has_code(&loan_receiver) {
+            return Err(Error::InvalidReceiver(ERC3156InvalidReceiver {
+                receiver,
+            }));
+        }
+        let loan_return = loan_receiver.on_flash_loan(
+            Call::new(),
+            msg::sender(),
+            token,
+            value,
+            fee,
+            data.to_vec().into(),
+        );
+        if loan_return.is_err() || loan_return.ok() != Some(RETURN_VALUE.into())
```

Comment:
Can we have `?` in a previous line and pattern match on Ok result? Tbh I do not like this if condition.

Reply:
I don't like it either, just left it for the refactor during final polishing.
Will change it immediately out of consideration for you :D